src/login/pam_systemd.c

```diff
@@ -374,37 +399,121 @@ _public_ PAM_EXTERN int pam_sm_open_session(
+                return PAM_SESSION_ERR;
+        }
+
+        r = sd_bus_message_open_container(m, 'a', "(sv)");
+        if (r < 0) {
+                pam_syslog(handle, LOG_ERR, "Failed to open message container: %s", strerror(-r));
+                return PAM_SYSTEM_ERR;
+        }
+
+        if (!isempty(memory_max)) {
+                if (streq(memory_max, "infinity")) {
+                        r = sd_bus_message_append(m, "(sv)", "MemoryMax", "t", (uint64_t)-1);
+                        if (r < 0) {
+                                pam_syslog(handle, LOG_ERR, "Failed to append to bus message: %s", strerror(-r));
+                                return PAM_SESSION_ERR;
+                        }
+                } else {
+                        r = parse_percent(memory_max);
+                        if (r >= 0) {
+                                r = sd_bus_message_append(m, "(sv)", "MemoryMaxScale", "u", (uint32_t) (((uint64_t) UINT32_MAX * r) / 100U));
+                                if (r < 0) {
+                                        pam_syslog(handle, LOG_ERR, "Failed to append to bus message: %s", strerror(-r));
+                                        return PAM_SESSION_ERR;
+                                }
+                        } else {
+                                r = parse_size(memory_max, 1024, &val);
+                                if (r >= 0) {
+                                        r = sd_bus_message_append(m, "(sv)", "MemoryMax", "t", val);
+                                        if (r < 0) {
+                                                pam_syslog(handle, LOG_ERR, "Failed to append to bus message: %s", strerror(-r));
+                                                return PAM_SESSION_ERR;
+                                        }
+                                } else
+                                        pam_syslog(handle, LOG_WARNING, "Failed to parse systemd.memory_max: %s, ignoring.", memory_max);
+                        }
+                }
+        }
+
+        if (!isempty(tasks_max) && !streq(tasks_max, "infinity")) {
+                /* No need to parse "infinity" here, it will be set unconditionally later in manager_start_scope() */
+                r = safe_atou64(tasks_max, &val);
+                if (r >= 0) {
+                        r = sd_bus_message_append(m, "(sv)", "TasksMax", "t", val);
+                        if (r < 0) {
+                                pam_syslog(handle, LOG_ERR, "Failed to append to bus message: %s", strerror(-r));
+                                return PAM_SESSION_ERR;
+                        }
+                } else
+                        pam_syslog(handle, LOG_WARNING, "Failed to parse systemd.tasks_max: %s, ignoring.", tasks_max);
+        }
+
+        r = append_session_cg_weight(m, cpu_weight, "CPUWeight");
+        if (r == 0)
+                pam_syslog(handle, LOG_WARNING, "Failed to parse systemd.cpu_weight: %s, ignoring.", cpu_weight);
```

Comment:
hmm, i see, r == 0 is also an error for this function.. hmm, this is very different from how we do this in our codebase usually, and i guess on unix in general, where errors usually are negative error codes, and only those... I'd really follow the same logic here, and minimize surprises to the reader...

I think it would be much nicer to just propagate all negative errors here, and move the error logging into append_session_cg_weight() instead, and just pass the pam handle in there...

Reply:
moving the error logging into append_session_cg_weight() would also cut down on the code, as the logging can be kept generic, and be reused for CPUWeight and IOWeight...